File: sources/global/util/src/main/java/com/tars/util/exceptions/ExceptionUtils.java
package com.tars.util.exceptions;

import org.slf4j.*;

import java.util.concurrent.*;
import java.util.function.*;

import javaslang.control.Try.*;
import util.*;

import static com.tars.util.validation.Validator.*;

/**
 * Utility class for exception handling and boilerplate elimination.
 *
 * note : logger with file appender must be configured for ExceptionUtils class.
 */
@Deprecated
public class ExceptionUtils {

  /**
   * Appropriate file appender shall be defined in logging configuration file.
   */
  public static final Logger log = LoggerFactory.getLogger("exceptions");

  // wtf methods

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf() {
    return reportAndDie(new WTFException());
  }

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf(String message) {
    return reportAndDie(new WTFException(message));
  }

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf(String message, Object... args) {
    return reportAndDie(
        new WTFException(
            String.format(message, args)
        )
    );
  }

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf(Throwable cause) {
    return reportAndDie(
        new WTFException(cause)
    );
  }

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf(Throwable cause, String message) {
    return reportAndDie(
        new WTFException(message, cause)
    );
  }

  /**
   * All overloaded wtf() methods are logging the provided data (if any) to a file and killing the process, these
   * represent use cases that shall never happen in production.
   *
   * P.S. WTF stands for (what a terrible failure)
   */
  public static <T> T wtf(Throwable cause, String message, Object... args) {
    return reportAndDie(new WTFException(
        String.format(message, args), cause
    ));
  }

  /**
   * Placeholder for work in progress.
   */
  public static <T> T notImplemented() {
    return reportAndDie(new WTFException("not implemented"));
  }

  // hooks

  private static final Option<Consumer<Throwable>> logCallback = Option.<Consumer<Throwable>>empty();

  /**
   * Sets callback to be executed when unrecoverable exception is about to kill the process. This can be used to notify
   * about critical failure or for graceful shutdown.
   */
  public static void onUnrecoverableFailure(Consumer<Throwable> task) {
    logCallback
        .map(consumer -> consumer.andThen(task))
        .ifNotPresentTake(task);
  }

  // logging to file

  /**
   * Logs an exception to general error log and returns without throwing it further.
   */
  public static void report(Throwable cause) {
    condition(notNull(cause));
    log.warn("unexpected exception", cause);
  }

  /**
   * Logs an exception to general error log and returns without throwing it further.
   */
  public static void report(Throwable cause, String message) {
    condition(notNull(cause) && notNullOrEmpty(message));
    log.warn(message, cause);
  }

  /**
   * If cause is checked exception wraps to RuntimeException
   */
  public static RuntimeException wrap(Throwable cause) {
    condition(notNull(cause));
    if (cause instanceof RuntimeException) {
      return (RuntimeException) cause;
    } else {
      return new RuntimeException(cause);
    }
  }

  /**
   * Logs an exception to general error log, if cause is checked exception wraps to RuntimeException and rethrows.
   */
  public static <T> T reportAndThrow(Throwable cause) {
    report(cause);
    throw wrap(cause);
  }

  /**
   * Logs an exception to general error log, if cause is checked exception wraps and returns a RuntimeException.
   */
  public static RuntimeException reportAndWrap(Throwable cause) {
    report(cause);
    return wrap(cause);
  }

  /**
   * Logs an exception to general error log with specified message, if cause is checked exception wraps and returns a
   * RuntimeException.
   */
  public static RuntimeException reportAndWrap(Throwable cause, String message) {
    report(cause, message);
    return wrap(cause);
  }

  private static <T> T reportAndDie(Throwable cause) {
    log.error("unrecoverable exception", cause);
    logCallback.ifPresent(throwableConsumer -> {
      throwableConsumer.accept(cause);
    });
    System.exit(-1); // kill process
    return null;
  }

  // execution utilities

  /**
   * Executes runnable, if any exception is thrown logs it and ignores.
   */
  public static void executeSilent(CheckedRunnable runnable) {
    try {
      runnable.run();
    } catch (Throwable cause) {
      report(cause);
    }
  }

  /**
   * Executes runnable, if any exception is thrown logs it, wraps to RuntimeException and rethrows.
   */
  public static void execute(CheckedRunnable runnable) {
    try {
      runnable.run();
    } catch (Throwable cause) {
      reportAndThrow(cause);
    }
  }

  /**
   * Executes runnable, if any exception is thrown logs it, executes callbacks if any, AND KILLS THE PROCESS.
   */
  public static void executeMandatory(CheckedRunnable runnable) {
    try {
      runnable.run();
    } catch (Throwable cause) {
      wtf(cause);
    }
  }

  // callable

  /**
   * Executes callable, wraps in Option, if any exception is thrown logs it and returns empty option.
   */
  public static <T> Option<T> executeAndGetSilent(Callable<T> callable) {
    try {
      return Option.ofNullable(callable.call());
    } catch (Throwable cause) {
      report(cause);
      return Option.<T>empty();
    }
  }

  /**
   * Executes callable, if callable returns null or any exception is thrown logs it, wraps to RuntimeException and
   * rethrows.
   */
  public static <T> T executeAndGet(Callable<T> callable) {
    try {
      T result = callable.call();

      if (result != null) {
        return result;
      } else {
        return reportAndThrow(new RuntimeException("callable returned null"));
      }
    } catch (Throwable cause) {
      return reportAndThrow(cause);
    }
  }

  /**
   * Executes callable, if callable returns null or any exception is thrown logs it, executes callbacks if any, AND
   * KILLS THE PROCESS.
   */
  public static <T> T executeAndGetMandatory(Callable<T> callable) {
    try {
      T result = callable.call();

      if (result != null) {
        return result;
      } else {
        throw new RuntimeException("callable returned null");
      }
    } catch (Throwable cause) {
      return wtf(cause);
    }
  }
}

File: sources/global/util/src/main/kotlin/util/exceptionUtils.kt
package util

import rx.subjects.PublishSubject
import util.global.logger
import util.global.whatever

/**
 * Utility class for exception handling and boilerplate elimination.
 *
 * All overloaded wtf() methods are logging the provided data (if any) to a file and kill the process, these
 * represent use cases that shall never happen in production. This is following fail fast principle.
 *
 * P.S. WTF stands for (what a terrible failure)
 *
 * note : logger with file appender must be configured for ExceptionUtils class.
 */
object exceptionUtils {

    /**
     * Appropriate file appender shall be defined in logging configuration file.
     */
    private val log by logger("exceptions")

    private val unrecoverableErrorStream = PublishSubject.create<Throwable>()

    /**
     * Register a listener for the unrecoverable errors.
     */
    fun onUnrecoverableFailure(listener: (Throwable) -> Unit) {
        unrecoverableErrorStream.subscribe({
            executeSilent { listener.invoke(it) }
        })
    }

    // wtf methods

    fun wtf() {
        reportAndKill(WTFException())
    }

    fun wtf(message: String) {
        reportAndKill(WTFException(message))
    }

    fun wtf(cause: Throwable) {
        reportAndKill(WTFException(cause))
    }

    fun wtf(cause: Throwable, message: String) {
        reportAndKill(WTFException(message, cause))
    }

    // logging to file

    /**
     * Logs an exception to general error log and returns without throwing it further.
     */
    fun report(cause: Throwable, message: String = "unexpected exception"): Throwable {
        log.info(message, cause)
        return cause
    }

    private fun reportAndKill(cause: Throwable) {
        // write log to a file
        log.error("unrecoverable error", cause)

        // acknowledge all the failure subscribers
        unrecoverableErrorStream.onNext(cause);

        // kill process
        System.exit(-1)
    }

    // runnable execution exception handling

    /**
     * Executes runnable, if any exception is thrown logs it and ignores.
     */
    fun executeSilent(block: () -> Unit) {
        try {
            block.invoke()
        } catch (ignored: Throwable) {
            report(ignored)
        }

    }

    /**
     * Executes runnable, if any exception is thrown logs it and rethrows.
     */
    fun execute(block: () -> Unit) {
        try {
            block.invoke()
        } catch (cause: Throwable) {
            throw report(cause)
        }

    }

    /**
     * Executes runnable, if any exception is thrown logs it, executes callbacks if any, AND KILLS THE PROCESS.
     */
    fun executeMandatory(block: () -> Unit) {
        try {
            block.invoke()
        } catch (cause: Throwable) {
            reportAndKill(WTFException(cause))
        }

    }

    // callable execution exception handling

    /**
     * Executes callable, wraps in Option, if any exception is thrown logs it and returns empty option.
     * Use case : failure is an expected execution flow.
     */
    fun <T> executeAndGetSilent(callable: () -> T): Option<T> {
        try {
            return Option.ofNullable(callable.invoke())
        } catch (cause: Throwable) {
            report(cause)
            return Option.empty<T>()
        }

    }

    /**
     * Executes callable, if callable returns null or any exception is thrown logs it, and rethrows.
     * Use case : failure is not expected but not fatal for the application.
     */
    fun <T> executeAndGet(callable: () -> T): T {
        try {
            val result = callable.invoke()

            if (result != null) {
                return result
            } else {
                throw report(RuntimeException("callable returned null"))
            }
        } catch (cause: Throwable) {
            throw report(cause)
        }
    }

    /**
     * Executes callable, if callable returns null or any exception is thrown logs it, executes callbacks if any, AND
     * KILLS THE PROCESS.
     * Use case : failure is not expected and is fatal for the application.
     */
    fun <T> executeAndGetMandatory(callable: () -> T): T {
        try {
            val result = callable.invoke()

            if (result != null) {
                return result
            } else {
                throw RuntimeException("callable returned null")
            }
        } catch (cause: Throwable) {
            return whatever { reportAndKill(WTFException(cause)) }
        }
    }

    // marker class

    /**
     * WTFException represents an unrecoverable error, a logical error that is not considered to ever happen in the system.
     */
    private class WTFException : RuntimeException {

        constructor() : super() {
        }

        constructor(message: String) : super(message) {
        }

        constructor(message: String, cause: Throwable) : super(message, cause) {
        }

        constructor(cause: Throwable) : super(cause) {
        }
    }
}

